rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -197,22 +196,19 @@ class RSpace[F[_]: Concurrent: ContextShift: Log: Metrics: Span, C, P, A, K](
       continuation: K,
       persist: Boolean,
       peeks: SortedSet[Int]
-  ): F[Consume] = syncF.delay {
-    eventLog.update(consumeRef +: _)
-    consumeRef
-  }
+  ): F[Consume] = for {
+      _ <- eventLog.update(log => consumeRef +: log)
+    } yield (consumeRef)
 
   protected override def logProduce(
       produceRef: Produce,
       channel: C,
       data: A,
       persist: Boolean
-  ): F[Produce] = syncF.delay {
-    eventLog.update(produceRef +: _)
-    if (!persist)
-      produceCounter.update(_.putAndIncrementCounter(produceRef))
-    produceRef
-  }
+  ): F[Produce] = for {
+      _ <- eventLog.update(log => produceRef +: log)
+      _ <- syncF.delay {if (!persist) produceCounter.update(_.putAndIncrementCounter(produceRef))}
```

Comment:
Expression inside `Sync[F].delay` has value of `F[Unit]` (precisely F[Any] because if..else are not the same type) and will not be executed. It's not part of _main_ execution (for...yield) expression, it's just a value created inside `delay` function.
For situations where value you want to wrap is already an F value there is `Sync[F].defer` function.

---

What is convenient here is `whenA` (and opposite `unlessA`, with `A` as part of Applicative interface) to conditionally execute given F value.
`_ <- produceCounter.update(_.putAndIncrementCounter(produceRef)).unlessA(persist)`

Reply:
In this place `produceCounter.update` return Unit (not F[Unit]). So, we can't call unlessA. I undestood your thought. I will use this in 3th task (replace VarSync to Ref for `produceCounter`).

As I understood right In this place we can rewrite all `logProduce` method like this:
```
eventLog.update(produceRef +: _).map(_ =>
        if (!persist) produceCounter.update(_.putAndIncrementCounter(produceRef))).as(produceRef)
```

Reply:
Oh yes, it's my mistake, I'm jumping ahead. Produce counter is updated in another PR.